language/bytecode_verifier/src/type_memory_safety.rs

```diff
@@ -845,17 +846,52 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
                 SignatureTokenView::new(self.module(), &struct_type).kind(self.type_formals());
 
                 let operand = self.stack.pop().unwrap();
-                if operand.signature == SignatureToken::Address {
-                    let nonce = self.get_nonce(&mut state);
-                    state.borrow_from_global_value(*idx, nonce.clone());
-                    self.stack.push(StackAbstractValue {
-                        signature: SignatureToken::MutableReference(Box::new(struct_type)),
-                        value: AbstractValue::Reference(nonce),
-                    });
-                    Ok(())
-                } else {
-                    Err(VMStaticViolation::BorrowGlobalTypeMismatchError(offset))
+                if operand.signature != SignatureToken::Address {
+                    return Err(VMStaticViolation::BorrowGlobalTypeMismatchError(offset));
+                }
+
+                let nonce = self.get_nonce(&mut state);
+                state.borrow_from_global_value(*idx, nonce.clone());
+                self.stack.push(StackAbstractValue {
+                    signature: SignatureToken::MutableReference(Box::new(struct_type)),
+                    value: AbstractValue::Reference(nonce),
+                });
+                Ok(())
+            }
+
+            Bytecode::ImmBorrowGlobal(idx, type_actuals_idx) => {
```

Comment:
I wonder if it is possible to factor out common checking code between MutBorrowGlobal and ImmBorrowGlobal.  Something like this was done for MutBorrowLoc and ImmBorrowLoc.

Reply:
I good idea, I was a bit too hasty in not doing so